peps/pep-0724.rst

```diff
@@ -0,0 +1,305 @@
+type checker to decide on how to implement. Future extensions to the type
+system may define this behavior though.
+
+
+Additional Examples
+===================
+
+Any
+---
+
+``Any`` is consistent [#isconsistent]_ with any other type, which means
+stricter semantics can be applied.
+
+.. code-block:: python
+
+    def is_a(x: Any) -> TypeGuard[A]:  # Stricter TypeGuard possible
+        return isinstance(x, A)
+
+    def test(x: A | B):
+        if is_a(x):
+            # x is of type A here
+        else:
+            # x is of type B here
+
+
+Backwards Compatibility
+=======================
+
+The new ``False`` case for a ``TypeGuard`` breaks backwards compatibility.
+
+.. code-block:: python
+
+    def is_int(val: int | str) -> TypeGuard[int]:
+        return isinstance(val, int)
+
+    def func(val: int | str):
+        if is_int(val):
+            # Type checker assumes int here
+        else:
+            # Type checker assumes str here
+
+A type checker will assume in the ``False`` case that the value is ``str``.
+
+This is a change in behavior from :pep:`647`.
+
+Although this is different behavior, existing code would likely have further
+checks for a ``str`` in the ``False`` case. Since the previous behavior was to
+assume ``str | int``, existing code would have had to further refine the type
+to get a type checker to behave the expected way.
+
+Something like so:
+
+.. code-block:: python
+
+    def is_int(val: int | str) -> TypeGuard[int]:
+        return isinstance(val, int)
+
+    def is_str(val: int | str) -> TypeGuard[int]:
+        return isinstance(val, int)
+
+    def func(val: int | str):
+        if is_int(val):
+            # Code does something with int here
+        else is_str(val):  # This check would have been necessary before
+            # Code does something with str here
+
+
+This means existing code should continue to work with the ``False`` narrowing.
+
+As a proof of this concept, we ran the `experimental Pyright changes`__
+against `mypy primer`__ to see if there were any differences in the output.
+Changing ``TypeGuard`` to be more strict had a small effect on the ``mypy
+primer``. Mostly indicating that some ``# type: ignore`` comments were no longer
+necessary.
+
+__ https://github.com/microsoft/pyright/pull/5832
+__ https://github.com/hauntsaninja/mypy_primer
+
+Breaking change
+---------------
+
+It is possible to create a broken ``TypeGuard`` now:
+
+.. code-block:: python
+
+    def is_positive_int(val: int | str) -> TypeGuard[int]:
+        return isinstance(val, int) and val > 0
+
+    def func(val: int | str):
+        if is_positive_int(val):
+            # Type checker assumes int here
+        else:
+            # Type checker now assumes str incorrectly here
+
+This case is now incorrect. However, we're confident that this is not a
+real world scenario. The mypy primer run didn't show any similar uses cases.
+Dynamic ``TypeGuard``\s are not a common pattern.
```

Comment:
You say you're confident that this form of type guard does not exist, but the next sentence says that dynamic guards are "not common", which seems a much weaker formulation. Please may you clarify the text?

Reply:
Sorry I'm not sure what else I'd say here. Mypy primer showed these dynamic type guards to not be used. We believe dynamic type guards aren't used because they don't make sense. You're not checking the type in a dynamic type guard, you're checking runtime behavior.

Reply:
Perhaps an additional example?

Reply:
@erictraut updated the text here. Hopefully that makes more sense.